web/modules/custom/joinup_federation/src/Plugin/pipeline/Step/AnalyzeIncomingEntities.php

```diff
@@ -0,0 +1,357 @@
+   *   The provenance helper service.
+   * @param \Drupal\rdf_schema_field_validation\SchemaFieldValidatorInterface $rdf_schema_field_validator
+   *   The RDF schema field validator service.
+   * @param \Drupal\joinup_federation\JoinupFederationHashGenerator $hash_generator
+   *   The hash generator service.
+   */
+  public function __construct(array $configuration, $plugin_id, array $plugin_definition, ConnectionInterface $sparql, EntityTypeManagerInterface $entity_type_manager, EntityFieldManagerInterface $entity_field_manager, ProvenanceHelperInterface $provenance_helper, SchemaFieldValidatorInterface $rdf_schema_field_validator, JoinupFederationHashGenerator $hash_generator) {
+    parent::__construct($configuration, $plugin_id, $plugin_definition, $sparql);
+    $this->entityTypeManager = $entity_type_manager;
+    $this->entityFieldManager = $entity_field_manager;
+    $this->provenanceHelper = $provenance_helper;
+    $this->rdfSchemaFieldValidator = $rdf_schema_field_validator;
+    $this->hashGenerator = $hash_generator;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition) {
+    return new static(
+      $configuration,
+      $plugin_id,
+      $plugin_definition,
+      $container->get('sparql_endpoint'),
+      $container->get('entity_type.manager'),
+      $container->get('entity_field.manager'),
+      $container->get('rdf_entity_provenance.provenance_helper'),
+      $container->get('rdf_schema_field_validation.schema_field_validator'),
+      $container->get('joinup_federation.hash_generator')
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function initBatchProcess() {
+    $incoming_ids = $this->getAllIncomingIds();
+    $this->setBatchValue('ids_to_process', $incoming_ids);
+    $this->setBatchValue('solution_ids', $this->getIncomingSolutionIds());
+    return ceil(count($incoming_ids) / self::BATCH_SIZE);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function batchProcessIsCompleted() {
+    return empty($this->getBatchValue('ids_to_process'));
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function execute() {
+    $ids = $this->extractNextSubset('ids_to_process', 50);
+    $solution_ids = array_intersect($ids, $this->getBatchValue('solution_ids'));
+
+    // Skip calculation of the same ids.
+    $hash_ids = array_diff($ids, $this->getEntityIdsWithHashes());
+    $this->setEntityHashes($this->hashGenerator->generateDataHash($hash_ids));
+
+    // Handle the solutions and their dependencies of this iteration.
+    $storage = $this->getRdfStorage();
+
+    /** @var \Drupal\rdf_entity\RdfInterface[] $entities */
+    $entities = $storage->loadMultiple($solution_ids, ['staging']);
+    foreach ($entities as $id => $entity) {
+      if ($entity->bundle() === 'solution') {
+        $this->buildSolutionDependencyTree($entity, $id);
```

Comment:
The thing with the `::buildSolutionDependency` is that, unlike the previous implementation, the dependency list is a flat list. What I mean is that if a solution references a release and that release references a distribution, the dependency tree array will not be
```
array => (
   solution => array (
      release => array(distribution)
   ),
)
```
but
```
array => (
  solution => (
     release, distribution,
  ),
)
```
That is why the id that is being passed in the method remains as the parent so that if it descends further into the recursion, all entities are listed directly under the solution and not as a structured array.